packages/ethash/src/index.ts

```diff
@@ -11,9 +11,127 @@ import {
 } from './util'
 // eslint-disable-next-line implicit-dependencies/no-implicit
 import type { LevelUp } from 'levelup'
-import type { Block, BlockHeader } from '@ethereumjs/block'
+import { Block, BlockHeader } from '@ethereumjs/block'
 const xor = require('buffer-xor')
 
+type Solution = {
+  mixHash: Buffer
+  nonce: Buffer
+}
+
+class Miner {
+  public blockHeader: BlockHeader
+  public block?: Block
+  public ethash: Ethash
+
+  public currentNonce: BN
+  public solution?: Solution
+
+  private headerHash?: Buffer
+  private stopMining: boolean
+
+  /**
+   * Create a Miner object
+   * @param mineObject - The object to mine on, either a `BlockHeader` or a `Block` object
+   * @param ethash - Ethash object to use for mining
+   */
+
+  constructor(mineObject: BlockHeader | Block, ethash: Ethash) {
+    if (mineObject.constructor === BlockHeader) {
+      this.blockHeader = mineObject
+    } else if (mineObject.constructor === Block) {
+      this.block = mineObject
+      this.blockHeader = mineObject.header
+    } else {
+      throw new Error('unsupported mineObject')
+    }
+    this.currentNonce = new BN(0)
+    this.ethash = ethash
+    this.stopMining = false
+  }
+
+  /**
+   * Stop the miner on the next iteration
+   */
+  stop() {
+    this.stopMining = true
+  }
+
+  /**
+   * Iterate `iterations` time over nonces, returns a `BlockHeader` or `Block` if a solution is found, `undefined` otherwise
+   * @param iterations - Number of iterations to iterate over. If `-1` is passed, the loop runs until a solution is found
+   * @returns - `undefined` if no solution was found within the iterations, or a `BlockHeader` or `Block`
+   *           with valid PoW based upon what was passed in the constructor
+   */
+  async mine(iterations: number = 0): Promise<undefined | BlockHeader | Block> {
+    const solution = await this.iterate(iterations)
+
+    if (solution) {
+      if (this.block) {
+        const data = this.block.toJSON()
+        data.header!.mixHash = '0x' + solution.mixHash.toString('hex')
+        data.header!.nonce = '0x' + solution.nonce.toString('hex')
+        return Block.fromBlockData(data)
```

Comment:
You should be able to just pass these in as Buffers, and I think you should specify the `common` so it is extended from the previous block.

```suggestion
        data.header!.mixHash = solution.mixHash
        data.header!.nonce = solution.nonce
        return Block.fromBlockData(data, { common: (this.block as any)._common })
```